cep-16.md

```diff
@@ -0,0 +1,250 @@
+# CEP for Sparse Repodata
+
+We propose a new "repodata" format that can be sparsely fetched. That means, generally, smaller fetches (only fetch what you need) and faster updates of existing repodata (only fetch what has changed).
+
+We also change the encoding from JSON to MSGPACK for faster decoding.
```

Comment:
Could this be considered a implementation detail? It imposes an additional dependency on implementors, but I think this should be mostly about the strategy of sharding, their path locations and the schema expecting. Maybe we can standardize things like `<path>/<filename>.<encoding>.<compression>`. Or at least standardize that a JSON counterpart is also available.

Reply:
Initially we used to hash the underlying data. Unfortunately that would not map well to the OCI registry. So we decided to use the hash of the compressed blob.

Reply:
And here are some basic numbers.

I generated the repodata shard index for conda-forges noarch subdir in both json and msgpack format. And I took the largest shard (`conda-forge-pinning`) for comparison as well.

```
| which               | size compressed | size uncompressed | deserialization time |
| ------------------- | --------------- | ----------------- | -------------------- |
| json shard index    | 705 KB          | 1404 KB           | 4.8 ms               |
| msgpack shard index | 666 KB (94%)    | 788 KB (56%)      | 3.2 ms (66%)         |
| json shard          | 298 KB          | 1926 KB           | 12.8 ms              |
| msgpack shard       | 313 KB (105%)   | 1567 KB (81%)     | 10.2 ms (83%)        |
```

In terms of size both JSON and msgpack files become roughly the same size. The biggest win seems to be in deserialization time which also takes up a significant amount of time in real-world examples.

Edit: If I change the format of the hashes in the shards from hex string to bytes, the msgpack files are always smaller (290 KB for the conda-forge-pinning case).

Reply:
This is the size of all shards of the noarch conda-forge channel (with the above edit applied):

```
|                      | size    |
| -------------------- | ------- |
| repodata.json.zst    | 13.5 MB |
| shards/*.json.zst    | 18   MB |
| shards/*.msgpack.zst | 17.1 MB |
```

Reply:
[This code plays with the codec and compression.](https://github.com/dholth/conda-test-data/blob/main/test_change_strategy.py#L190-L193)

Reply:
Have you drawn any conclusions from this that we should incorporate @dholth ?